src/env.rs

```diff
@@ -0,0 +1,491 @@
+use regex::{Captures, Regex};
+use std::borrow::Borrow;
+use std::collections::HashMap;
+use std::env;
+use std::hash::Hash;
+use std::ops::Index;
+
+lazy_static! {
+    static ref ENV_VAR_REGEX: Regex = Regex::new(r"(\$[\w\?\-#!\$_@\*]*)").unwrap();
+    static ref PARTIAL_BRACKET_ENV_VAR_REGEX: Regex =
+        Regex::new(r"(\$\{([\w\?\-#!\$_@\*]*)\}?)").unwrap();
+    static ref BRACKET_ENV_VAR_REGEX: Regex = Regex::new(r"(\$\{([\w\?\-#!\$_@\*]+)\})").unwrap();
+}
+
+type Key = String;
+type Value = String;
+type Map = HashMap<Key, Value>;
+
+/// Env encapsulates environment variables and their manipulation.
+pub struct Env {
+    env: Map,
+}
+
+impl Env {
+    pub fn new() -> Env {
+        Env {
+            env: env::vars().collect(),
+        }
+    }
+
+    pub fn insert(&mut self, key: Key, value: Value) {
+        self.env.insert(key, value);
+    }
+
+    pub fn remove<S>(&mut self, key: &S)
+    where
+        S: ?Sized + Hash + Eq,
+        Key: Borrow<S>,
+    {
+        self.env.remove(key);
+    }
+
+    pub fn get<S>(&self, key: &S) -> Option<&Value>
+    where
+        S: ?Sized + Hash + Eq,
+        Key: Borrow<S>,
+    {
+        self.env.get(key)
+    }
+
+    pub fn contains_key<S>(&self, key: &S) -> bool
+    where
+        S: ?Sized + Hash + Eq,
+        Key: Borrow<S>,
+    {
+        self.env.contains_key(key)
+    }
+
+    /// Append value to value at key but only if current value doesn't already contain input value.
+    /** If key doesn't exist then an empty string entry will be created. */
+    pub fn append<S>(&mut self, key: &S, value: Value)
+    where
+        S: ?Sized + Hash + Eq + ToString,
+        Key: Borrow<S>,
+    {
+        if !self.env.contains_key(key) {
+            self.env.insert(key.to_string(), "".to_string());
+        }
+        let old_value = self.env[&key].clone();
+        if !old_value.contains(&value) {
+            self.env.insert(key.to_string(), old_value + &value);
+        }
+    }
+
+    /// Replace value with another value for value at key but only if current value already is
+    /// contained.
+    /** If key doesn't exist then an empty string entry will be created. */
+    pub fn replace<S>(&mut self, key: &S, old_value: Value, new_value: Value)
+    where
+        S: ?Sized + Hash + Eq + ToString,
+        Key: Borrow<S>,
+    {
+        if !self.env.contains_key(key) {
+            self.env.insert(key.to_string(), "".to_string());
+        }
+        let value = self.env[key].clone();
+        if value.contains(&old_value) {
+            self.env
+                .insert(key.to_string(), value.replace(&old_value, &new_value));
+        }
+    }
+
+    pub fn print(&self) {
```

Comment:
Can we instead implement the `std::fmt::Display` trait?

Reply:
Good idea!